File: python_kata/src/graphs/dijkstra.py
def dijkstra(graph, src, dest):
    unvisited = set()
    for node in graph:
        unvisited.add(node)
    distances = {}
    for node in graph:
        if node == src:
            distances[node] = 0
        else:
            distances[node] = float("inf")
    return dijkstra_r(graph, src, dest, unvisited, distances, {})


def dijkstra_r(graph, src, dest, unvisited, distances, predecessors):
    if src == dest:
        path = []
        pred = dest
        while pred != None:
            path.append(pred)
            pred = predecessors.get(pred, None)
        path.reverse()
        return path
    min_dist = float("inf")
    min_neighbor = src
    for neighbor in graph[src]:
        print(f"neighbor: ", neighbor)
        if neighbor in unvisited:
            distance_so_far = distances[src]
            distance_to_neighbor = graph[src][neighbor]
            total_distance_to_neighbor = distance_so_far + distance_to_neighbor
            if total_distance_to_neighbor < distances.get(neighbor):
                distances[neighbor] = total_distance_to_neighbor
                predecessors[neighbor] = src
            if total_distance_to_neighbor < min_dist:
                min_dist = total_distance_to_neighbor
                min_neighbor = neighbor
    unvisited.remove(src)
    return dijkstra_r(graph, min_neighbor, dest, unvisited, distances, predecessors)

File: typescript-kata/src/graphs/dijkstra.ts
/**
   * NOT OPTIMIZED Dijkstra algoritm implementation.
   * O(V^2 + E)
   * @param  {Set<T>} unvisited  We track what are 
   * @param  {String} sep   A period separating the initials.
   * @param  {String} trail A period ending the initials.
   * @param  {String} hyph  A hypen separating double names.
   * @return {String}       Properly formatted initials.
*/
export class Dijkstra {
    constructor() {

    }

    dijkstra(graph: Map<string, Map<string, number>>, source: string, destination: string): string[] {
        const unvisited = new Set<string>;
        const distances = new Map<string, number>();
        const predecessors = new Map<string, string>();

        for (const node of graph.keys()) {
            unvisited.add(node);
            // console.log(unvisited)
        }

        for (const node of graph.keys()) {
            if (node === source) {
                distances.set(source, 0)
            } else {
                distances.set(node, Infinity);
            }
        }



        return this.walk(graph, source, destination, unvisited, distances, predecessors);

    }

    private walk(graph: Map<string, Map<string, number>>, source: string, destination: string, unvisited: Set<string>, distances: Map<string, number>, predecessors: Map<string, string>): string[] {
        if (source === destination) {
            return this.getPath(predecessors, destination);
        }

        let minDistance = Infinity;
        let minNeighbor = source;

        const neighbors: string[] = [];
        graph.get(source)?.forEach((value, key) => {
            neighbors.push(key)
        });

        for (const neighbor of neighbors as string[]) {

            if (unvisited.has(neighbor)) {
                let distanceSoFar = distances.get(source) as number;
                let distanceToNeighbor = graph.get(source)?.get(neighbor) as number;

                const totalDistanceToNeighbor = distanceSoFar + distanceToNeighbor;
                const distanceToNeighborFromDistances = distances.get(neighbor) as number;

                if (totalDistanceToNeighbor < distanceToNeighborFromDistances) {
                    distances.set(neighbor, totalDistanceToNeighbor);
                    predecessors.set(neighbor, source);
                }

                if (totalDistanceToNeighbor < minDistance) {
                    // console.log("CHUUJ", source)
                    minDistance = totalDistanceToNeighbor;
                    minNeighbor = neighbor
                    // console.log("Chuuja min: ", minNeighbor)
                }

            }
        }
        unvisited.delete(source);

        return this.walk(graph, minNeighbor, destination, unvisited, distances, predecessors);

    }



    private getMinDistanceNode(distances: Map<string, number>, unvisited: Set<string>): string | undefined {
        let out: string = "";
        let minVal: number = Infinity;

        if (unvisited.size === 0) {
            return undefined;
        }

        for (const node of unvisited) {
            if (distances.get(node) as number < minVal) {
                out = node;
                minVal = distances.get(node) as number;
            }
        }

        return out
    }

    private getPath(predecessors: Map<string, string>, destination: string) {
        const out: string[] = [];
        out.push(destination);

        while (true) {
            const next = predecessors.get(out[out.length - 1]);

            if (next === undefined) {
                break;
            }

            out.push(next);
        }

        return out.reverse();
    }

}
